### ISSUE TITLE
Race condition in oauth2_provider.oauth2_validators.OAuth2Validator#get_original_scopes

### ISSUE THREAD

**Issue by pySilver** [evonove/django-oauth-toolkit#314]
I belive there is a race condition in oauth2_provider.oauth2_validators.OAuth2Validator#get_original_scopes

Imagine two requests (very possible) that tries to obtain new access token using refresh_token grant type. Once first succeed, the other one contains invalid refrence to access token which results to AccessToken.DoesNotExists exception. This is caused by the fact that first request did remove _old_ access token and created new one.

To avoid this either keep ID of Access Token the same or reselect Access Token

**Comment by flc** [comment]
I think I ran into a similar (or the same) problem. If I request a new access token with a valid refresh token in parallel I get `DoesNotExist: AccessToken matching query does not exist.` errors from the `AccessToken.objects.get(id=self.access_token.id).revoke()` call.

Sounds like wrapping RefreshToken.revoke with a `transaction.atomic` decorator would be a good idea.

**Comment by pySilver** [comment]
Actually there is something even more interesting happening. It took us few days and a lot of logging to find out simple root of our problems. First of all it has nothing to do with this lib directly but rather your server setup (uwsgi & nginx). We are using short-living access tokens (1H) as advertised by specs and this leads to a situation when all clients tries to update their tokens often.

And here is the problem. While issuing new refresh token at: ``OAuth2Validator#save_bearer_token`` its also deletes *old refresh token* used in original request. But you have no idea if client received response. It may (and in our case thats was true) be killed by nginx or uwsgi timeout. So client never received a new token, but his refresh token is already removed, so he cannot just simply replay failed request. 

Solution:
# Tune your server setup (check your logs for IOError)
# Increase Access Token expiration (leads to a larger attack window)
# Do not delete refresh tokens immediately, but instead schedule it for deletion in a next few minutes, so client can replay failed request (also affects security)

#206 #266

**Comment by pySilver** [comment]
However this lib by default regenerates refresh token, however its not requried by specs:

```
The authorization server MAY issue a new refresh token, in which case
   the client MUST discard the old refresh token and replace it with the
   new refresh token.  The authorization server MAY revoke the old
   refresh token after issuing a new refresh token to the client.  If a
   new refresh token is issued, the refresh token scope MUST be
   identical to that of the refresh token included by the client in the
   request.
````
https://tools.ietf.org/html/draft-ietf-oauth-v2-31#page-43

**Comment by synasius** [comment]
Should be fixed in #389
